src/sentry/integrations/source_code_management/repository.py

```diff
@@ -0,0 +1,200 @@
+from sentry.integrations.services.repository import RpcRepository, repository_service
+from sentry.models.identity import Identity
+from sentry.models.integrations.integration import Integration
+from sentry.models.repository import Repository
+from sentry.shared_integrations.client.base import BaseApiResponseX
+from sentry.shared_integrations.exceptions import ApiError, IntegrationError
+
+
+class RepositoryIntegration(ABC):
+    repo_search = False
+    model: RpcIntegration | Integration
+
+    # @abstractmethod
+    @property
+    def codeowners_locations(self) -> Sequence[str] | None:
+        """
+        A list of possible locations for the CODEOWNERS file.
+        """
+        return None
+
+    @abstractmethod
+    def get_client(self) -> RepositoryClient:
+        """Returns the client for the integration. The client must be a subclass of RepositoryClient."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def source_url_matches(self, url: str) -> bool:
+        """Checks if the url matches the integration's source url."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def format_source_url(self, repo: Repository, filepath: str, branch: str | None) -> str:
+        """Formats the source code url used for stack trace linking."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def extract_branch_from_source_url(self, repo: Repository, url: str) -> str:
+        """Extracts the branch from the source code url."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def extract_source_path_from_source_url(self, repo: Repository, url: str) -> str:
+        """Extracts the source path from the source code url."""
+        raise NotImplementedError
+
+    def check_file(self, repo: Repository, filepath: str, branch: str | None = None) -> str | None:
+        """
+        Calls the client's `check_file` method to see if the file exists.
+        Returns the link to the file if it's exists, otherwise return `None`.
+
+        So far only GitHub, GitLab and VSTS have this implemented, all of which give
+        use back 404s. If for some reason an integration gives back a different
+        status code, this method could be overwritten.
+
+        repo: Repository (object)
+        filepath: file from the stacktrace (string)
+        branch: commitsha or default_branch (string)
+        """
+        filepath = filepath.lstrip("/")
+        try:
+            client = self.get_client()
+        except (Identity.DoesNotExist, IntegrationError):
+            sentry_sdk.capture_exception()
+            return None
+        try:
+            response = client.check_file(repo, filepath, branch)
+            if response is None:
+                return None
+        except IdentityNotValid:
+            return None
+        except ApiError as e:
+            if e.code != 404:
+                sentry_sdk.capture_exception()
+                raise
```

Comment:
These might be good areas to gather some metrics around failures in case we hit issues with the specific integrations/providers.